### DISCUSSION TITLE
Is it possible to self host outline without a publicly accessible URL?

### DISCUSSION THREAD

**Discussion by Hommus** [outline/outline#2138] [Self Hosting] [answered]
As stated in the title, Is it possible to self host outline without a publicly accessible URL?

Most of the self-hosted applications I use are accessible through a static IP and port, from a NAS. I've tried changing the PORT environment variable on the env.outline, and everything boots up successfully but there is no website present at the PORT (7000).

**Comment by tommoor** [comment, marked as answer]
Honestly I don't know, noone has tried that I'm aware of but in a local environment it's pretty normal to use `/etc/hosts` to map a domain to an IP address only on your machine – you could do similar.

**Reply by Hommus** [reply]
I'll give it a shot, cheers

**Reply by auziabbruzzesi** [reply]
Did you have any luck with that? I'm having the same issue

**Reply by Hommus** [reply]
It ended up working without the need for editing /etc/hosts. I followed [this guide](https://github.com/chsasank/outline-wiki-docker-compose) to setup a local version. Then, to get it working on a server:
- I updated the docker-compose.yml (IP, PORT, Volumes)
- Copied the contents of data/nginx to the server (I think this is what made it work for me)
- Moved the environment variables to the docker-compose.yml instead of having .env files (This was unnecessary, but a requirement of my server)

One caveat is that you will not be able to sign in using google as they require a publicly accessible URL.

Updated docker-compose.yml (IP: 192.168.1.50, PORT: 7000):
```
version: "3"
services:
  redis:
    image: redis
    restart: always
  postgres:
    image: postgres
    environment:
      POSTGRES_USER: user
      POSTGRES_PASSWORD: pass
      POSTGRES_DB: outline
    volumes:
      - /outline/pgdata:/var/lib/postgresql/data
    restart: always
  minio:
    image: minio/minio
    volumes:
      - /outline/minio_root:/minio_root:z
      - /outline/certs:/root/.minio/certs:z
    command: "minio server /minio_root"
    environment:
      - MINIO_ACCESS_KEY=
      - MINIO_SECRET_KEY=
      - MINIO_BROWSER=off
    restart: always
  outline:
    image: outlinewiki/outline:latest
    command: sh -c "yarn sequelize:migrate --env production-ssl-disabled && yarn start"
    environment:
      - SECRET_KEY=
      - UTILS_SECRET=
      - DATABASE_URL=postgres://user:pass@postgres:5432/outline
      - DATABASE_URL_TEST=postgres://user:pass@postgres:5432/outline-test
      - REDIS_URL=redis://redis:6379
      - URL=http://192.168.1.50:7000
      - PORT=3000
      - CDN_URL=http://192.168.1.50:7000
      - FORCE_HTTPS=false
      - ENABLE_UPDATES=true
      - DEBUG=cache,presenters,events,emails,mailer,utils,multiplayer,server,services
      - GOOGLE_CLIENT_ID=
      - GOOGLE_CLIENT_SECRET=
      - AWS_ACCESS_KEY_ID=
      - AWS_SECRET_ACCESS_KEY=
      - AWS_REGION=xx-xxxx-x
      - AWS_S3_UPLOAD_BUCKET_URL=http://192.168.1.50:7000
      - AWS_S3_UPLOAD_BUCKET_NAME=outline-bucket
      - AWS_S3_UPLOAD_MAX_SIZE=26214400
      - AWS_S3_FORCE_PATH_STYLE=true
      - AWS_S3_ACL=private
      - DEFAULT_LANGUAGE=en_US
      - HOST=192.168.1.50
      - HTTP_PORT=3000
      - PGSSLMODE=disable
    restart: always
    depends_on:
      - postgres
      - redis
      - minio
  nginx:
    image: nginx
    ports:
      - 7000:80
      - 4443:443
    volumes:
      - /outline/nginx/:/etc/nginx/conf.d/:z
      - /outline/certs/:/etc/ssl/certs/:z
    restart: always
    depends_on:
      - outline
```

**Reply by auziabbruzzesi** [reply]
Thank you! This is exactly what I ended up doing -- I worked on it into the wee hours. evidently I should have just gone to bed and read your response in the morning. Cheers.

**Reply by Hommus** [reply]
Glad you got it working!

**Reply by bym0** [reply]
Hey @Paulie-P,
i am just curious, does google really require a publicly accesible URL? Because it's just a redirect with a token. Most of the OAuth Provider only need a URL like http://localhost:3000/oauth/callback (if you would host it on your machine). I dont think you are really dependent on a publicly accessible URl.

**Reply by Hommus** [reply]
Google will accept http://localhost:3000/oauth/callback as a valid redirect URI. If you want to use an IP address however, you will be met with the following error:
> Invalid origin: Must end with a public top-level domain (such as .com or .org).

There is a thread here for more discussion on the matter: 
https://stackoverflow.com/questions/14238665/can-a-public-ip-address-be-used-as-google-oauth-redirect-uri

In my case, we have several computers connected to a local NAS which we access using an IP address. We could setup the NAS to be accessible by a public top-level domain, but we decided to use gitea as our authentication method instead.
